### PR TITLE
Fix a regression whereby payment details are not saved from the AdditionalPayment form

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
I discovered the Payment.create api is not respecting a bunch of parameters - since this is now called from the AdditionalPaymentForm they are not being saved.

payment_instrument_id, card_type_id, pan_truncation & other field were not being passed through - resulting in incorrect payment saves

Before
----------------------------------------
Add a pending contribution &  then submit a credit card payment using the additional payment form

Screen Shot 2019-10-18 at 7 24 48 PM

After
----------------------------------------
Screen Shot 2019-10-18 at 7 41 03 PM

Technical Details
----------------------------------------
This looks worse than it is because I also had to pull in my upstream param renaming commit https://github.com/civicrm/civicrm-core/pull/15535

Comments
----------------------------------------
@monishdeb @jitendrapurohit can one of you check

### CODE REVIEW

File: api/v3/utils.php [hunk 1]
```diff
@@ -390,6 +390,9 @@ function _civicrm_api3_get_BAO($name) {
     // not the other cache info like search results (which could in fact be in Redis or another cache engine)
     $name = 'PrevNextCache';
   }
+  if ($name === 'Payment') {
+    $name = 'FinancialTrxn';
+  }
```
Comment (mattwire) [inline api/v3/utils.php]: @eileenmcnaughton Is this change required for this PR? As I understand it Payment does not map directly to FinancialTrxn but wraps the financial entities.

File: api/v3/Payment.php [hunk 2]
```diff
@@ -187,6 +197,119 @@ function _civicrm_api3_payment_create_spec(&$params) {
...
+      'size' => 25,
+      'where' => 'civicrm_financial_trxn.order_reference',
+      'table_name' => 'civicrm_financial_trxn',
+      'entity' => 'FinancialTrxn',
+      'bao' => 'CRM_Financial_DAO_FinancialTrxn',
+      'localizable' => 0,
+      'html' => [
+        'type' => 'Text',
+      ],
+    ],
```
Comment (mattwire) [inline api/v3/Payment.php]: @eileenmcnaughton `order_reference` was only added in 5.20 - will this cause problems on 5.19 as the column won't exist in the DB?

File: api/v3/Payment.php [hunk 3]
```diff
@@ -168,9 +171,16 @@ function _civicrm_api3_payment_create_spec(&$params) {
       'type' => CRM_Utils_Type::T_FLOAT,
     ],
     'payment_processor_id' => [
-      'title' => ts('Payment Processor ID'),
+      'name' => 'payment_processor_id',
       'type' => CRM_Utils_Type::T_INT,
-      'description' => ts('Payment processor ID - required for payment processor payments'),
+      'title' => ts('Payment Processor'),
+      'description' => ts('Payment Processor for this financial transaction'),
```
Comment (mattwire) [inline api/v3/Payment.php]: Can we use `Payment processor for this payment`.  As the payment API will be creating one or more financial transactions depending on the action.

File: CRM/Financial/BAO/Payment.php [hunk 4]
```diff
@@ -59,27 +59,36 @@ public static function create($params) {
 
     $isPaymentCompletesContribution = self::isPaymentCompletesContribution($params['contribution_id'], $params['total_amount']);
 
+    $whiteList = ['check_number', 'payment_processor_id', 'fee_amount', 'total_amount', 'contribution_id', 'net_amount', 'card_type_id', 'pan_truncation', 'trxn_result_code', 'payment_instrument_id', 'order_reference', 'trxn_id'];
```
Comment (mattwire) [inline CRM/Financial/BAO/Payment.php]: @eileenmcnaughton See comment on `order_reference` which is only available from 5.20 - though I don't think it will cause problems in this list.

**Review by mattwire — DISMISSED** [review]
@eileenmcnaughton I've added a couple of comments.  The main concern is around `order_reference` which did not exist in 5.19.
Is there a minimum version of this PR without the variable rename that could be merged for 5.19 and then do the rest for 5.20?
As this is modifying the internals of `CRM_Financial_BAO_Payment::create` it has the potential to break badly if we've missed anything!

File: api/v3/utils.php (hunk 1 above)
Comment (eileenmcnaughton) [reply api/v3/utils.php]: @mattwire yes - otherwise the pseudoconstant doesn't work.

Payment is a psuedo-api that represents one specific type of FinancialTrxn - hence the presence of 'is_payment' as a fixed parameter when creating a financial_trxn from the payment api.

Ie it's a subset of the FinancialTrxn entity

File: api/v3/Payment.php (hunk 2 above)
Comment (eileenmcnaughton) [reply api/v3/Payment.php]: Yeah - I think it's the cause of the test fails - I just pulled it out again

File: api/v3/Payment.php (hunk 3 above)
Comment (eileenmcnaughton) [reply api/v3/Payment.php]: Yes - good point

File: CRM/Financial/BAO/Payment.php (hunk 4 above)
Comment (eileenmcnaughton) [reply CRM/Financial/BAO/Payment.php]: yep - removed

**Comment by eileenmcnaughton** [comment]
@mattwire I've fixed the order_reference part. To be honest I started working on 5.20 & then moved back to 5.19 once I realised I'd hit a regression. 

The variable rename is 'safe' & hopefully my comments on the other PR address your concerns. If so we can merge that to sync this with current master & rebase this to be cleaner

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@magnolia61

**Comment by mattwire** [comment]
@eileenmcnaughton Please rebase and I'll take another look

**Comment by eileenmcnaughton** [comment]
@mattwire done - also FYI - the reason I moved the param gathering outside the IF is that the same params are actually required if total_amount < 0 - that code is for historical reasons in it's own function but over a few refactors it would start to share that paymentParams array

**Comment by eileenmcnaughton** [comment]
Hopefully that sorts the test. Along with https://github.com/civicrm/civicrm-core/pull/15561 we should aim for 5.18

**Comment by mattwire** [comment]
test this please

**Comment by jitendrapurohit** [comment]
I've tested this on contribution, event payments and confirm that the params are correctly set after the submission. The payment.create api call can also set these values in the record.

Failing test is SyntaxConformanceTest.php::testLimit for Payment Entity. Should this entity be added to `static::toBeSkipped_getlimit()` as well?

**Comment by seamuslee001** [comment]
merging as per the collective reviews
